File: packages/print-service-client/src/pos-presale-ticket.ts
/**
 * Ticket de preventa POS → agente KaiPrinters (`type: "pos-presale-ticket"`).
 */

import type { PosSaleTicketCompany } from "./pos-sale-ticket";

export const POS_PRESALE_TICKET_PAYLOAD_VERSION = 1;

export type PosPresaleTicketLine = {
  productName: string;
  variantName?: string | null;
  quantity: number;
  total: number;
};

export type PosPresaleTicketPayload = {
  version: number;
  code: string;
  qrPayload: string;
  issuedAt: string;
  company: PosSaleTicketCompany;
  branchName?: string | null;
  pointOfSaleName?: string | null;
  lines: PosPresaleTicketLine[];
  total: number;
};

export type PosPresaleTicketPrintExtras = {
  filename: string;
  documentType?: string;
  internalFolio?: string;
  sourceApp?: string;
  purpose?: string;
};

File: kai-printers-desktop/src-tauri/src/pos_presale_ticket.rs
//! Ticket de preventa POS → JSON (`pos-presale-ticket`).

use crate::pos_sale_ticket_pdf::TicketCompany;
use anyhow::{Context, Result};
use serde::Deserialize;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresaleLine {
    pub product_name: String,
    pub variant_name: Option<String>,
    pub quantity: f64,
    pub total: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PosPresaleTicket {
    #[serde(default)]
    #[allow(dead_code)]
    pub version: i32,
    pub code: String,
    pub qr_payload: String,
    pub issued_at: String,
    pub company: TicketCompany,
    pub branch_name: Option<String>,
    pub point_of_sale_name: Option<String>,
    #[serde(default)]
    pub lines: Vec<PresaleLine>,
    pub total: f64,
    #[serde(default)]
    pub operator_name: Option<String>,
}

pub fn parse_pos_presale_ticket_from_value(value: &serde_json::Value) -> Result<PosPresaleTicket> {
    serde_json::from_value(value.clone()).context("parse pos-presale-ticket")
}

pub fn write_pos_presale_ticket_escpos_from_value(
    dir: &PathBuf,
    value: &serde_json::Value,
) -> Result<PathBuf> {
    crate::pos_presale_ticket_escpos::write_pos_presale_ticket_escpos_from_value(dir, value)
}
